module: Control and parse gem5's statistics log file

qperret opened the pull request:
Gem5's statistics log file contains plenty of interesting information
that are not exposed so far. This module enables control and parsing of
the statistics file by:
 - configuring periodic dumps of statistics;
 - marking Regions of Interest (ROIs);
 - and extracting values of specific fields during the ROIs.

setrofim commented on devlib/module/gem5stats.py:
```diff
@@ -0,0 +1,132 @@
+                            records[key][roi].append(dump[key])
+        return records
+
+    def _iter_dump(self, stats_file):
```
This does not need to be a method, as it does not rely on any of the module's internal state. Making this into a stand-alone function would make it easier to reuse. I would also move this out into `devlib/utils/gem5.py`

setrofim commented on devlib/module/gem5stats.py:
```diff
@@ -0,0 +1,132 @@
+            for dump in self._iter_dump(stats_file):
+                for roi in xrange(GEM5STATS_ROI_NUMBER):
+                    # Save records only when ROIs are ON
+                    if int(dump['ROI::{}'.format(roi)])==1:
```
Spaces around `==`

setrofim replied:
Is `'ROI::{}'.format(roi)` guaranteed to be in the dump, even if ROI are not used? If not, this should be made more robust.

qperret replied:
Good question ... The ROI fields have always been present during my runs, they were just equal to 0 all the time if no ROI was used. Maybe @Sticklyman1936 could confirm there are no special cases/configs where the ROI fields are not dumped ?

setrofim commented on devlib/module/gem5stats.py:
```diff
@@ -0,0 +1,132 @@
+        self._current_origin = 0
+        self._stats_file_path = os.path.join(target.platform.gem5_out_dir,
+                                            'stats.txt') 
+        self.logger = logging.getLogger('Gem5Stats')
```
This is unnecessary as the logger is created by the super.

setrofim commented on devlib/module/gem5stats.py:
```diff
@@ -0,0 +1,132 @@
+import devlib
+from devlib.module import Module
+from devlib.platform import Platform
+
```
Two lines between includes and globals

setrofim commented on devlib/module/gem5stats.py:
```diff
@@ -0,0 +1,132 @@
+GEM5STATS_DUMP_HEAD = '---------- Begin Simulation Statistics ----------'
+GEM5STATS_DUMP_TAIL = '---------- End Simulation Statistics   ----------'
+GEM5STATS_ROI_NUMBER = 8
+
```
Two lines between globals and class declarations

setrofim commented on devlib/module/gem5stats.py:
```diff
@@ -0,0 +1,132 @@
+
+    @staticmethod
+    def probe(target):
+       return isinstance(target.platform, devlib.platform.gem5.Gem5SimulationPlatform)
```
Import `Gem5SimulationPlatform` at the top and just use the short name here.

qperret commented:
I moved the dump iterator into utils.gem5 as requested and I added a ROI booking system. The idea is that several clients (different instruments for example) might use this module simultaneously and they should not share ROIs without knowing it. To solve the issue I'm proposing to hide the ROI numbers from clients which will use labels instead.
